src/codegen.cpp

```diff
@@ -2511,8 +2512,16 @@ static bool emit_builtin_call(jl_codectx_t &ctx, jl_cgval_t *ret, jl_value_t *f,
                 else if (!isboxed && jl_is_uniontype(ety)) {
                     Type *AT = ArrayType::get(IntegerType::get(jl_LLVMContext, 8 * al), (elsz + al - 1) / al);
                     Value *data = emit_bitcast(ctx, emit_arrayptr(ctx, ary, ary_ex), AT->getPointerTo());
-                    // isbits union selector bytes are stored directly after the last array element
-                    Value *selidx = emit_arraylen_prim(ctx, ary);
+                    // isbits union selector bytes are stored after a->maxsize
+                    Value *ndims = (nd == -1 ? emit_arrayndims(ctx, ary) : ConstantInt::get(T_int16, nd));
+                    Value *is_vector = ctx.builder.CreateICmpEQ(ndims, ConstantInt::get(T_int16, 1));
+                    Value *offset = emit_arrayoffset(ctx, ary);
+                    Value *selidx_v = ctx.builder.CreateSub(emit_vectormaxsize(ctx, ary), ctx.builder.CreateZExt(offset, T_size));
+                    Value *selidx_m = emit_arraylen(ctx, ary);
+                    Value *selidx = ctx.builder.CreateSelect(is_vector, selidx_v, selidx_m);
+                    Value *elsize = emit_arrayelsize(ctx, ary);
```

Comment:
I added these three lines since the type tag first element should be `(a->maxsize - a->offset) * a->elsize + a->offset`. Not sure if we need to add a `CreateZExt` for `elsize` or `offset` here though? (we have one above when we subtract maxsize and offset). I tried w/ and I got segfaults either way.